File: Exam_Review/Practice_Problems/DP.c
#include <stdlib.h>
#include <limits.h>

int min(const int a, const int b) {
    return (a < b) ? a : b;
}

// Q2: Minimum number of coins for a target value
int min_num_coins(int *denominations, int num_denom, int target) {
    int *DP = (int *)malloc(sizeof(int) * (target + 1));
    for (int i = 0; i <= target; i++) {
        DP[i] = INT_MAX;
    }
    DP[0] = 0;

    for (int i = 1; i <= target; i++) {
        for (int j = 0; j < num_denom; j++) {
            if (i - denominations[j] >= 0 && DP[i - denominations[j]] != INT_MAX) {
                DP[i] = min(DP[i], DP[i - denominations[j]] + 1);
            }
        }
    }

    int result = DP[target];
    free(DP);
    return result == INT_MAX ? -1 : result;
}

// Q3: List min coins for all values up to target
int *list_min_coins(int *denominations, int num_denom, int target) {
    int *DP = (int *)malloc(sizeof(int) * target);
    for (int i = 0; i < target; i++) {
        DP[i] = min_num_coins(denominations, num_denom, i + 1);
    }
    return DP;
}

// Q4: Return actual coins used to make up the target value
int *coins_req(int *denominations, int num_denom, int target, int *res_size) {
    int *DP = (int *)malloc(sizeof(int) * (target + 1));
    int *prev = (int *)malloc(sizeof(int) * (target + 1));

    for (int i = 0; i <= target; i++) {
        DP[i] = INT_MAX;
        prev[i] = -1;
    }
    DP[0] = 0;

    for (int i = 1; i <= target; i++) {
        for (int j = 0; j < num_denom; j++) {
            int coin = denominations[j];
            if (i - coin >= 0 && DP[i - coin] != INT_MAX && DP[i - coin] + 1 < DP[i]) {
                DP[i] = DP[i - coin] + 1;
                prev[i] = coin;
            }
        }
    }

    if (DP[target] == INT_MAX) {
        *res_size = 0;
        free(DP);
        free(prev);
        return NULL;  // No solution
    }

    *res_size = DP[target];
    int *res = (int *)malloc(sizeof(int) * (*res_size));
    int curr = target, idx = 0;

    while (curr > 0) {
        int coin = prev[curr];
        res[idx++] = coin;
        curr -= coin;
    }

    free(DP);
    free(prev);
    return res;
}

// Q. 5

#include <string.h>

int canBeSegmented(char *s, char **wordDict, int num_words) {
    if (s[0] == '\0') {
        return 1; // Base case: successfully segmented entire string
    }

    for (int i = 0; i < num_words; i++) {
        int word_len = strlen(wordDict[i]);
        if (strncmp(s, wordDict[i], word_len) == 0) {
            if (canBeSegmented(s + word_len, wordDict, num_words)) {
                return 1;
            }
        }
    }

    return 0; // No valid segmentation found
}

File: Exam_Review/Practice_Problems/DP.py
# Lab 8 Questions -----------------------------------

# Q. 2

def min_coins(val, denominations):
    dp = [float('inf')] * (val + 1)
    dp[0] = 0  # Base case: 0 coins needed to make 0

    for i in range(1, val + 1):
        for coin in denominations:
            if i - coin >= 0:
                dp[i] = min(dp[i], dp[i - coin] + 1)

    return dp[val] if dp[val] != float('inf') else -1  # -1 if not possible

# Q. 3

def min_coin_list(val, denominations):
    dp = [float('inf')] * (val + 1)
    dp[0] = 0  # Base case: 0 coins needed to make 0

    for i in range(1, val + 1):
        for coin in denominations:
            if i - coin >= 0:
                dp[i] = min(dp[i], dp[i - coin] + 1)

    return dp

# Q. 4

def coins_req(val, denominations):
    res = []
    while val > 0:
        for d in denominations:
            if val - d >= 0 and min_coins(val - d, denominations) == min_coins(val, denominations) - 1:
                res.append(d)
                val -= d
                break  # Move to the next value after choosing a coin
    return res

# Q. 5

def canBeSegmented(s, WordDict):
    if s == "":
        return True
    for word in WordDict:
        if s.startswith(word):
            if canBeSegmented(s[len(word):]):
                return True
    return False
